File: app/src/main/java/nz/org/cacophony/sidekick/db/RoomDatabase.java
package nz.org.cacophony.sidekick.db;


import android.content.Context;
import android.os.AsyncTask;

import androidx.annotation.NonNull;
import androidx.room.Database;
import androidx.room.Room;
import androidx.sqlite.db.SupportSQLiteDatabase;

@Database(entities = {Recording.class, Event.class}, version = 4, exportSchema = false)
public abstract class RoomDatabase extends androidx.room.RoomDatabase {

    // marking the instance as volatile to ensure atomic access to the variable
    private static volatile RoomDatabase INSTANCE;
    private static androidx.room.RoomDatabase.Callback sRoomDatabaseCallback = new androidx.room.RoomDatabase.Callback() {

        @Override
        public void onOpen(@NonNull SupportSQLiteDatabase db) {
            super.onOpen(db);
            new PopulateDbAsync(INSTANCE).execute();
        }
    };

    public static RoomDatabase getDatabase(final Context context) {

        if (INSTANCE == null) {
            synchronized (RoomDatabase.class) {
                if (INSTANCE == null) {
                    INSTANCE = Room.databaseBuilder(context.getApplicationContext(),
                            RoomDatabase.class, "recording_database")
                            .addMigrations(Migrations.getMIGRATION_3_4())
                            .addCallback(sRoomDatabaseCallback)
                            .build();
                }
            }
        }
        return INSTANCE;
    }

    public abstract RecordingDao recordingDao();

    public abstract EventDao eventDao();

    private static class PopulateDbAsync extends AsyncTask<Void, Void, Void> {

        private final RecordingDao mDao;

        PopulateDbAsync(RoomDatabase db) {
            mDao = db.recordingDao();
        }

        @Override
        protected Void doInBackground(final Void... params) {
            // Start the app with a clean database every time.
            // Not needed if you only populate on creation.
            //mDao.deleteAll();
            return null;
        }
    }
}

File: app/src/main/java/nz/org/cacophony/sidekick/db/RoomDatabase.kt
package nz.org.cacophony.sidekick.db

import android.content.Context
import android.util.Log
import androidx.room.Database
import androidx.room.Room
import nz.org.cacophony.sidekick.TAG
import nz.org.cacophony.sidekick.db.Migrations.Companion.MIGRATION_3_4
import nz.org.cacophony.sidekick.db.Migrations.Companion.MIGRATION_4_5
import java.io.File

@Database(entities = [Recording::class, Event::class], version = 5, exportSchema = false)
abstract class RoomDatabase : androidx.room.RoomDatabase() {

    abstract fun recordingDao(): RecordingDao
    abstract fun eventDao(): EventDao

    fun clearData() {
        val recordings = recordingDao().getAllRecordings()
        for (rec in recordings) {
            Log.i(TAG, rec.recordingPath)
            File(rec.recordingPath).delete()
            recordingDao().deleteRecording(rec.id)
        }
        val events = eventDao().getAllEvents()
        for (e in events) {
            eventDao().delete(e)
        }
    }

    companion object {

        @Volatile
        private var INSTANCE: RoomDatabase? = null

        fun getDatabase(context: Context): RoomDatabase? {
            if (INSTANCE == null) {
                synchronized(RoomDatabase::class.java) {
                    if (INSTANCE == null) {
                        INSTANCE = Room.databaseBuilder(context.applicationContext,
                                RoomDatabase::class.java, "recording_database")
                                .addMigrations(MIGRATION_3_4, MIGRATION_4_5)
                                .build()
                    }
                }
            }
            return INSTANCE
        }
    }
}
